### PR TITLE
Refactor grants methods

### PR DESCRIPTION
**Pull request by PedramNavid**
This is a pretty big refactor of the existing grants methods, and combines @blake-enyart and Sumit Rai's work on future grants as well. 

The biggest change here is to the `snowflake_grants` module. Here's a brief overview of the changes:
* rename the class method `check_grant_to_role` to `is_granted_privilege` to help clarify what it does
* break apart the massive 400+ line `generate_grant_roles` class method into a 50 line method, with all the functionality split into smaller private methods.
* break apart the even more massive `generate_table_and_view_grants` into smaller private methods
* add future grants capabilities 

A few things to note here: the private methods are still quite large, complex, and have too many arguments. There's a lot of duplicated code, but at least breaking them apart shows how conflated some of the concerns have become. It should also make adding unit tests easier as we can now test individual grant generations.

While all the tests passed, I still need to look over coverage of the individual units and see if there is good coverage.

### CODE REVIEW

File: src/permifrost/core/permissions/utils/snowflake_grants.py [hunk 1]
```diff
@@ -1259,12 +1382,15 @@ def generate_table_and_view_grants(  # noqa
...
-            all_grant_views = read_grant_views_full + write_grant_views_full
             view_split = granted_view.split(".")
             database_name = view_split[0]
 
             # For future grants at the database level
-            if len(view_split) == 2:
+            # TODO: Not sure what is happenign with table split here, it
+            # may posibly be unbound
+            if len(view_split) == 2 or (
+                len(table_split) == 3 and table_split[1] == "*"
```
Comment (PedramNavid) [inline src/permifrost/core/permissions/utils/snowflake_grants.py:1392]: @blake-enyart I think the table_split piece here is from your revokes. I'm a little confused here, it looks like `table_split` comes from the `for granted_table in list(..)` iteration on line 1324. Why would the `for granted_view in list...` iteration reference table_split? If there are not `granted_tables` then `table_split` will be undefined as well

File: src/permifrost/core/permissions/utils/snowflake_grants.py [hunk 2]
```diff
@@ -1281,6 +1407,11 @@ def generate_table_and_view_grants(  # noqa
                 # Don't revoke on privileges on databases not defined in spec.
                 continue
             elif granted_view == future_view and future_view not in all_grant_views:
+                print("*" * 100)
```
Comment (PedramNavid) [inline src/permifrost/core/permissions/utils/snowflake_grants.py:1410]: @blake-enyart should we be printing here? I don't see any other print commands within the utils function.

File: src/permifrost/core/permissions/utils/snowflake_grants.py [hunk 3]
```diff
@@ -928,6 +926,55 @@ def generate_table_and_view_grants(  # noqa
 
             fetched_schemas = conn.full_schema_list(f"{name_parts[0]}.{name_parts[1]}")
 
+            # For future grants at the database level
+            future_database_table = "{database}.<table>".format(database=name_parts[0])
```
Comment (PedramNavid) [inline src/permifrost/core/permissions/utils/snowflake_grants.py:930]: @blake-enyart FYI I refactored the method and moved some of yoru code here into the private `_generate_table_read_grants` method

File: src/permifrost/core/permissions/utils/snowflake_grants.py [hunk 4]
```diff
@@ -928,6 +926,55 @@ def generate_table_and_view_grants(  # noqa
...
+            future_database_view = "{database}.<view>".format(database=name_parts[0])
+
+            table_already_granted = False
+            view_already_granted = False
+
+            if self.is_granted_privilege(
+                role, read_privileges, "table", future_database_table
+            ):
+                table_already_granted = True
+
```
Comment (PedramNavid) [inline src/permifrost/core/permissions/utils/snowflake_grants.py:940]: I think we can simplify this statement to:
```table_already_granted = self.is_granted_privilege(....)```

File: src/permifrost/core/permissions/utils/snowflake_grants.py [hunk 5]
```diff
@@ -928,6 +926,55 @@ def generate_table_and_view_grants(  # noqa
...
+                            grouping_name=SnowflakeConnector.snowflaky(name_parts[0]),
+                            role=SnowflakeConnector.snowflaky(role),
+                        ),
+                    }
+                )
+
+            if self.is_granted_privilege(
+                role, read_privileges, "view", future_database_view
+            ):
+                view_already_granted = True
```
Comment (PedramNavid) [inline src/permifrost/core/permissions/utils/snowflake_grants.py:960]: same comment as above

File: src/permifrost/core/permissions/utils/snowflake_grants.py (hunk 2 above)
Comment (blake-enyart) [reply src/permifrost/core/permissions/utils/snowflake_grants.py:1410]: Great catch, I think this snuck through from the MR over on GitLab and needs to be removed. Will do that shortly

File: src/permifrost/core/permissions/utils/snowflake_grants.py (hunk 3 above)
Comment (blake-enyart) [reply src/permifrost/core/permissions/utils/snowflake_grants.py:930]: Sounds good

File: src/permifrost/core/permissions/utils/snowflake_grants.py (hunk 4 above)
Comment (blake-enyart) [reply src/permifrost/core/permissions/utils/snowflake_grants.py:940]: Agreed and did this for both the view and table commands

**Comment by blake-enyart** [comment]
@PedramNavid, I branched off of this `refactor-grants` branch and merged the changes you left on this PR into this as well. I hope that was okay.

Overall, this looks great, but I think the test harness was very light on this section. I'm going to try and do some work on that before we meet up tomorrow to prove out some other scenarios for grants generation.

File: src/permifrost/core/permissions/utils/snowflake_grants.py [hunk 6]
```diff
@@ -62,63 +62,47 @@ def check_grant_to_role(
...
-    ) -> List[Dict]:
-        """
-        Generate the GRANT statements for both roles and users.
+        return False
 
-        entity_type: "users" or "roles"
-        entity: the name of the entity (e.g. "yannis" or "reporter")
-        config: the subtree for the entity as specified in the spec
-        all_entities: all roles defined in spec
+    def _generate_member_lists(self, config) -> Tuple[List, List]:
```
[the remark this answers is not in the document]
Comment (PedramNavid) [reply src/permifrost/core/permissions/utils/snowflake_grants.py:71]: Great idea, let me do this for all the functions as best I can

**Pull request merged**